src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs

```diff
@@ -0,0 +1,346 @@
+
+        if (_partitionCount > maxPartitionCount)
+        {
+            _partitionCount = maxPartitionCount;
+            _maxBatchSize = (int)(recordCount / _partitionCount);
+        }
+
+        // Calculating estimated pool margin at about 5% of total working size. The working set size fluctuate a bit so
+        // this is to reduce allocation.
+        long estimatedPoolMargin = (long)(((double)recordCount / 128) * 0.05);
+        ObjectPool<CompactStack<Job>.Node> jobPool = new DefaultObjectPool<CompactStack<Job>.Node>(
+            new CompactStack<Job>.ObjectPoolPolicy(128), (int)estimatedPoolMargin);
+
+        _currentPointer = 0;
+        _queuedJobs = 0;
+        _activeJobs = 0;
+        _targetCurrentItems = _partitionCount * _maxBatchSize;
+
+        // This need to be very small
+        _partitions = new CompactStack<Job>[_partitionCount];
+        for (int i = 0; i < _partitionCount; i++)
+        {
+            _partitions[i] = new CompactStack<Job>(jobPool);
+        }
+    }
+
+    // Determine the locality of the key. I guess if you use paprika or something, you'd need to modify this.
+    int CalculatePartitionIdx(ValueKeccak key)
+    {
+        uint number = BinaryPrimitives.ReadUInt32BigEndian(key.Span);
+        return (int)(number * (ulong)_partitionCount / uint.MaxValue);
+    }
+
+    public void Start(
+        ValueKeccak root,
+        TrieVisitContext trieVisitContext)
+    {
+        // Start with the root
+        SmallTrieVisitContext rootContext = new(trieVisitContext);
+        _partitions[CalculatePartitionIdx(root)].Push(new Job(root, rootContext));
+        _activeJobs = 1;
+        _queuedJobs = 1;
+
+        try
+        {
+            Task[]? tasks = Enumerable.Range(0, trieVisitContext.MaxDegreeOfParallelism)
+                .Select((_) => Task.Run(BatchedThread))
+                .ToArray();
+
+            Task.WhenAll(tasks).Wait();
+        }
+        catch (Exception)
+        {
+            _failed = true;
+            throw;
+        }
+    }
+
+    ArrayPoolList<(TrieNode, SmallTrieVisitContext)>? GetNextBatch()
+    {
+        CompactStack<Job>? theStack;
+        do
+        {
+            long partitionIdx = Interlocked.Increment(ref _currentPointer);
+            if (partitionIdx == _partitionCount)
+            {
+                Interlocked.Add(ref _currentPointer, -_partitionCount);
+
+                GC.Collect(); // Simulate GC collect of standard visitor
+            }
+            partitionIdx %= _partitionCount;
+
+            if (_activeJobs == 0 || _failed)
+            {
+                // Its finished
+                return null;
+            }
+
+            if (_queuedJobs == 0)
+            {
+                // Just a small timeout to prevent threads from loading CPU
+                // Note, there are other threads also going through the stacks, so its fine to have this high.
```

Comment:
btw have you checked how this full pruning method affects block processing? That would be something interesting from validator's perspective
If not, we should check it during next experiments (cc: @kamilchodola )

Reply:
It quite a lot worst. Likely due to higher rate of processing. Seems to load CPU quite heavily.